### PR TITLE
Convert /exchangeRates endpoint to use bulk queries to data providers

### PR DESCRIPTION
**Pull request by peachbits**
This commit overhauls the rates server to use bulk data endpoints.

- The following historical data endpoints are now batching requests
  coinmarketcap
  currencyconverter
  openexchangerates (new backup fiat data source)

- The following latest data endpoints are now batching requests
  coinmarketcap
  coincap
  nomics (new)
  compound (new)

Coincap historical doesn't not support batching so those are still single requests. The zero and fallback rates are also techincally bulk now as well.

Other changes:
- Use async-lock to prevent multiple writes to the same document.
- Each requested rate obj can include a date.
- Partially added Edge-preferred unique IDs. This takes the map of currency codes and unique IDs from a data source and ensures the default currencies in Edge the queried values in case of duplicates. Only using it for coincap and it does not dynamically refresh the original list of 2000.
- Currency bridge uses bridge currencies to connect to two existing rates to find a third unknown rate (ie. requested BCH_JPY, can return a rate with BCH_USD and JPY_USD). This is synchronous and doesn't create new queries to data providers.
- Crypto data providers now only query rates against USD. Fiat data providers will query against USD.
- Currencies pegged to an underlying asset (ie. WETH, AUSDC, etc) use the underlying asset price.
- Updated /exchangeRate endpoint to use new request code.

### CODE REVIEW

File: src/providers/coinMarketCapBasic.ts [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    headers: {
+      'X-CMC_PRO_API_KEY': config.coinMarketCapCurrentApiKey
+    },
+    json: true
+  }
+  if (codesWanted.length > 0)
+    try {
+      const codes = codesWanted.join(',')
+      const response = await fetch(
+        `https://pro-api.coinmarketcap.com/v1/cryptocurrency/quotes/latest?symbol=${codes}&skip_invalid=true`,
```
Comment (peachbits) [inline src/providers/coinMarketCapBasic.ts:53]: no literal

File: src/providers/coincap.ts [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    // Query
+    for (const date in datesAndCodesWanted) {
+      if (datesAndCodesWanted[date].length === 0) continue
+      rates[date] = {}
+
+      if (date === currentTime) {
+        // Latest data endpoint accepts bulk requests
+        const codes = createUniqueIdString(datesAndCodesWanted[date])
+        if (codes === '') continue
+        const url = `https://api.coincap.io/v2/assets?ids=${codes}`
```
Comment (peachbits) [inline src/providers/coincap.ts:61]: no literal

File: src/providers/coincap.ts [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+    const fromCurrency = checkConstantCode(pair.currency_pair.split('_')[0])
+    if (fiatCurrencyCodes[fromCurrency] == null) {
+      if (datesAndCodesWanted[pair.date] == null) {
+        datesAndCodesWanted[pair.date] = []
+      }
+      datesAndCodesWanted[pair.date].push(fromCurrency)
+    }
+
+    // Query
+    for (const date in datesAndCodesWanted) {
```
Comment (peachbits) [inline src/providers/coincap.ts:53]: promise.all

File: src/providers/coincap.ts [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+          // Add to return object
+          json.data.forEach(obj => {
+            rates[date][`${obj.symbol}_USD`] = obj.priceUsd
+          })
+        } catch (e) {
+          log(`No coincapCurrent quote: ${JSON.stringify(e)}`)
+        }
+      } else {
+        // Historical data endpoint is limited to one currency at a time
+        for (const code of datesAndCodesWanted[date]) {
```
Comment (peachbits) [inline src/providers/coincap.ts:81]: promise.all

File: src/providers/coincap.ts [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+          const timestamp = Date.parse(date)
+          const id = createUniqueIdString([code])
+          if (id === '') continue
+          try {
+            const response = await fetch(
+              `https://api.coincap.io/v2/assets/${id}/history?interval=m1&start=${timestamp}&end=${timestamp +
```
Comment (peachbits) [inline src/providers/coincap.ts:87]: no literal

File: src/providers/coincap.ts [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+      }
+      datesAndCodesWanted[pair.date].push(fromCurrency)
+    }
+
+    // Query
+    for (const date in datesAndCodesWanted) {
+      if (datesAndCodesWanted[date].length === 0) continue
+      rates[date] = {}
+
+      if (date === currentTime) {
```
Comment (peachbits) [inline src/providers/coincap.ts:57]: separate current and historical functions

File: src/providers/currencyConverter.ts [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+      fiatCurrencyCodes[toCurrency] === true &&
+      fromCurrency !== 'USD' &&
+      datesAndCodesWanted[pair.date].indexOf(`USD_${toCurrency}`) === -1
+    ) {
+      datesAndCodesWanted[pair.date].push(`USD_${toCurrency}`)
+    }
+  }
+
+  // Query
+  for (const date in datesAndCodesWanted) {
```
Comment (peachbits) [inline src/providers/currencyConverter.ts:51]: promise.all

File: src/providers/currencyConverter.ts [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+        throw new Error(
+          `currencyConverter returned with status: ${JSON.stringify(
+            status ?? response.status
+          )} and error: ${JSON.stringify(error)}`
+        )
+      }
+
+      // Create return object
+      rates[date] = {}
+      for (const pair of Object.keys(results)) {
```
Comment (peachbits) [inline src/providers/currencyConverter.ts:79]: standardize

Comment (peachbits) [reply src/providers/currencyConverter.ts:79]: Object.assign(rates, Object.keys(results).reduce((rates, pair)=> ({...rates, [pair]:results[pair].val[justDate].toString()}), {}))

File: src/providers/openExchangeRates.ts [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+    if (
+      fiatCurrencyCodes[toCurrency] === true &&
+      datesAndCodesWanted[pair.date].indexOf(toCurrency) === -1
+    ) {
+      datesAndCodesWanted[pair.date].push(toCurrency)
+    }
+  }
+
+  // Query
+  for (const date in datesAndCodesWanted) {
```
Comment (peachbits) [inline src/providers/openExchangeRates.ts:47]: promise.all

File: src/config.ts [hunk 10]
```diff
@@ -68,14 +68,29 @@ export const asConfig = asObject({
...
+    asString,
+    'https://openexchangerates.org'
+  ),
   nomicsApiKey: asOptional(asString, NOMICS_API_KEY),
+  nomicsBaseUrl: asOptional(asString, 'https://api.nomics.com'),
+  coincapBaseUrl: asOptional(asString, 'https://api.coincap.io'),
+  compoundBaseUrl: asOptional(
+    asString,
+    'https://api.compound.finance/api/v2/ctoken'
+  ),
```
Comment (thehobbit85) [inline src/config.ts]: Create a config param like this:
```
providers: {
   currencyConverter: {
       uri: *****,
       apiKey: ****
   }
 }
 ```

File: src/providers/openExchangeRates.ts [hunk 11]
```diff
@@ -44,35 +45,44 @@ const openExchangeRates = async (
...
         )
-        throw new Error(
-          `openExchangeRates returned with status: ${JSON.stringify(
-            status ?? response.status
-          )} and error: ${JSON.stringify(response.statusText)}`
-        )
-      }
+      })
+    })
+  )
```
Comment (thehobbit85) [inline src/providers/openExchangeRates.ts]: This is not how you should convert this into Promise.all.
Let's have a sidebar and I'll show you how it should be done.

**Review by thehobbit85 — CHANGES_REQUESTED** [review]
See comments

File: src/providers/coincap.ts [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+  codes: string[],
+  log: Function
+): Promise<ProviderResponse> => {
+  const rates = { [date]: {} }
+  const codeString = createUniqueIdString(codes)
+  if (codeString === '') return rates
+  const url = `${uri}/v2/assets?ids=${codes}`
+  try {
+    const response = await fetch(url, OPTIONS)
+    const json = asCoincapCurrentResponse(await response.json())
```
Comment (peachbits) [inline src/providers/coincap.ts:48]: normalize response and create shared function

**Pull request merged**